### PR TITLE
Feature: entity and dto suffix (#8647)

### PR DESCRIPTION
**Pull request by mselerin**
-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

PR for #8647

### CODE REVIEW

File: generators/app/index.js [hunk 1]
```diff
@@ -105,6 +105,20 @@ module.exports = class extends BaseGenerator {
             defaults: 'jhi'
         });
 
+        // This adds support for a `--domain-suffix` flag
+        this.option('domain-suffix', {
```
Comment (gmarziou) [inline generators/app/index.js]: "domain suffix" does not really talk to me, if you google these terms you'll find DNS related content and this is what also what it means to me. Based on description I would rather call it "entity suffix". What do others think?

Comment (DanielFran) [reply generators/app/index.js]: Yes, 'entity suffix' sounds better to me also.

Comment (mselerin) [reply generators/app/index.js]: I called it 'domain' to match the package where the class stands. But I'm ok with 'Entity Suffix', it makes more sense.

I'll change this tomorrow.

File: generators/app/index.js [hunk 2]
```diff
@@ -145,6 +159,13 @@ module.exports = class extends BaseGenerator {
         this.skipCheckLengthOfIdentifier = this.configOptions.skipCheckLengthOfIdentifier =
             this.options['skip-check-length-of-identifier'] || this.config.get('skipCheckLengthOfIdentifier');
         this.jhiPrefix = this.configOptions.jhiPrefix = _.camelCase(this.config.get('jhiPrefix') || this.options['jhi-prefix']);
+
+        this.entitySuffix = this.configOptions.entitySuffix =
+            this.config.get('entitySuffix') === false ? '' : this.config.get('entitySuffix') || this.options['entity-suffix'];
```
Comment (deepu105) [inline generators/app/index.js]: You don't need the conditional, jist short circuiting is enough

File: generators/entity-server/index.js [hunk 3]
```diff
@@ -43,6 +43,10 @@ module.exports = class extends BaseGenerator {
         } else {
             useBlueprint = false;
         }
+
+        // Expose 'asEntity' and 'asDto' functions for templating
+        this.asEntity = name => name + this.entitySuffix;
```
Comment (deepu105) [inline generators/entity-server/index.js]: Please move these methods to generator-base.js

File: generators/app/index.js (hunk 2 above)
Comment (mselerin) [reply generators/app/index.js]: I need the `this.config.get('entitySuffix') === false` because when the parameters come from the jdl (`entitySuffix false`), it's a real false value. So we need to test it and convert it to an empty string.

If the value stays to `false`, the entity name would be `Foofalse`.

I could have done the test in the `asEntity` function but the separation of concerns principle tells me to not. ;-)

Unless I misunderstood what you mean...

File: generators/entity-server/index.js (hunk 3 above)
Comment (mselerin) [reply generators/entity-server/index.js]: Ok. I wasn't sure if the generator-base just for those functions was the right place but if you tell so, I'm 100% ok with that.

File: generators/app/index.js (hunk 2 above)
Comment (deepu105) [reply generators/app/index.js]: Ok understood

File: generators/entity-server/index.js (hunk 3 above)
Comment (deepu105) [reply generators/entity-server/index.js]: Yes its the right place

Comment (mselerin) [reply generators/entity-server/index.js]: Just to let you know...

If I move `asEntity` and `asDto` to generator-base, I still need to define these inside a constructor (which does not yet exists in generator-base). Otherwise, the methods can't be used inside the templates.

I can do something like this : 

```javascript
module.exports = class extends PrivateBase {
  constructor(args, opts) {
        super(args, opts);
        
        // Expose asEntity and asDto to templates
        this['asEntity'] = this.asEntity;
        this['asDto'] = this.asDto;
    }

    /**
     * Construct the entity name by appending the entity suffix.
     * @param {String} name entity name
     */
    asEntity(name) {
        return name + this.entitySuffix;
    }

    /**
     * Construct the entity's dto name by appending the dto suffix.
     * @param {String} name entity name
     */
    asDto(name) {
        return name + this.dtoSuffix;
    }
};
```

But the constructor feel a little hacky... :-(

Comment (mselerin) [reply generators/entity-server/index.js]: OK I got it !
Everything is fine until a template call include(). For example :
```
<%- include('../../common/save_template', {viaService: viaService, returnDirectly: true}); -%>
```
When the function are defined inside the constructor (see above), I don't need to pass any additional values to the include function.

To fix this, I need to send more parameters to include, like this : 
```
<%- include('../../common/save_template', {asEntity; asEntity, asDto: asDto, viaService: viaService, returnDirectly: true}); -%>
```

Go with that solution ?

Comment (deepu105) [reply generators/entity-server/index.js]: hmmm, thats weird. I don't see how `this.asEntity = this['asEntity'];` is different from just declaring a class member method. Technically its the same. Anyway i'm ok with passing these to includes if needed

File: generators/app/index.js [hunk 4]
```diff
@@ -145,6 +159,13 @@ module.exports = class extends BaseGenerator {
...
+
+        this.dtoSuffix = this.configOptions.dtoSuffix =
```
Comment (deepu105) [inline generators/app/index.js]: @mselerin where are this being saved to config?

File: generators/entity-server/templates/src/main/java/package/service/impl/EntityServiceImpl.java.ejs [hunk 5]
```diff
@@ -78,14 +78,14 @@ import static org.elasticsearch.index.query.QueryBuilders.*;
...
- * Service Implementation for managing <%= entityClass %>.
+ * Service Implementation for managing <%= asEntity(entityClass) %>.
  */
 @Service<% if (databaseType === 'sql') { %>
 @Transactional<% } %>
 public class <%= serviceClassName %><% if (service === 'serviceImpl') { %> implements <%= entityClass %>Service<% } %> {
 
     private final Logger log = LoggerFactory.getLogger(<%= serviceClassName %>.class);
-<%- include('../../common/inject_template', {viaService: viaService, constructorName: serviceClassName, queryService: false}); -%>
+<%- include('../../common/inject_template', {asEntity: asEntity, asDto: asDto, viaService: viaService, constructorName: serviceClassName, queryService: false}); -%>
```
Comment (deepu105) [inline generators/entity-server/templates/src/main/java/package/service/impl/EntityServiceImpl.java.ejs]: since we are on node 8+ can you check if the shortHand syntax works here?
```suggestion
<%- include('../../common/inject_template', {asEntity, asDto, viaService, constructorName: serviceClassName, queryService: false}); -%>
```

File: generators/app/index.js (hunk 4 above)
Comment (mselerin) [reply generators/app/index.js]: Same files, lines 380-381 in the `saveConfig()` function : 
```javascript
            saveConfig() {
                const config = {
                    jhipsterVersion: packagejs.version,
                    applicationType: this.applicationType,
                    baseName: this.baseName,
                    testFrameworks: this.testFrameworks,
                    jhiPrefix: this.jhiPrefix,
                    entitySuffix: this.entitySuffix,
                    dtoSuffix: this.dtoSuffix,
                    skipCheckLengthOfIdentifier: this.skipCheckLengthOfIdentifier,
                    otherModules: this.otherModules,
                    enableTranslation: this.enableTranslation,
                    clientPackageManager: this.clientPackageManager
                };
```

File: generators/entity-server/templates/src/main/java/package/service/impl/EntityServiceImpl.java.ejs (hunk 5 above)
Comment (mselerin) [reply generators/entity-server/templates/src/main/java/package/service/impl/EntityServiceImpl.java.ejs]: OK this seems to work well. I have made the change everywhere.

File: generators/app/index.js (hunk 4 above)
Comment (deepu105) [reply generators/app/index.js]: Sorry missed it somehow

**Comment by mselerin** [comment]
Documentation done !
https://github.com/jhipster/jhipster.github.io/pull/654

**Comment by MathieuAA** [comment]
In the docs, it says that the suffixes can take two types for values:
  - a string (for an actual, non-empty suffix)
  - false (for no suffix)

Why not just an empty string?

**Comment by mselerin** [comment]
> In the docs, it says that the suffixes can take two types for values:
> 
> * a string (for an actual, non-empty suffix)
> * false (for no suffix)
> 
> Why not just an empty string?

Because it seems not possible to specify an empty string inside the JDL. But if there's a way, I'll take it :-)

**Comment by MathieuAA** [comment]
Okay, there may be a mismatch between what's generated in JCore (in the JSON entity files) and what's generated by the generator. I'll update JCore to only output empty strings.

**Comment by DanielFran** [comment]
@mselerin do not forget to update PR in accordance with the changes from Mathieu.

**Comment by MathieuAA** [comment]
@DanielFran There's nothing to change here (already done in JCore). That's why I asked the question: I was afraid of a potential value/type mismatch between the generated JSON files from JCore and here, but the issue only was in JCore, so nothing to do :)

**Comment by mselerin** [comment]
BTW, the suffixes are not present in the JSON entity files. These parameters are application-wide, therefore only present inside the `.yo-rc.json` file.

And to be complete, an empty value inside `.yo-rc.json` is stored like this : 

```json
"dtoSuffix": ""
```

The `false` value is only needed inside a JDL file.

**Comment by DanielFran** [comment]
@MathieuAA I was refering for the documentário 😎

**Comment by MathieuAA** [comment]
Ahhhhh!Okay :)

**Comment by deepu105** [comment]
@jhipster/developers I'm tight on time can someone test and merge this?

**Comment by pascalgrimaud** [comment]
@deepu105 : we have our JHipster day at Ippon, this tuesday. I think our priority will be to test and merge all pending PR ; then try to reduce the number of opened tickets.

I'll try to help for this

File: generators/app/index.js (hunk 2 above)
Comment (MathieuAA) [reply generators/app/index.js]: Not anymore. Deepu's point still stands. The JDL doesn't export the suffixes as false (entity & dto)

File: generators/app/index.js [hunk 6]
```diff
@@ -145,6 +159,13 @@ module.exports = class extends BaseGenerator {
...
+            this.config.get('dtoSuffix') === false ? '' : this.config.get('dtoSuffix') || this.options['dto-suffix'];
```
Comment (MathieuAA) [inline generators/app/index.js]: Same as my previous comment

File: generators/entity/index.js [hunk 7]
```diff
@@ -223,6 +223,18 @@ module.exports = class extends BaseGenerator {
                     this.log(chalk.green(`\nFound the ${context.filename} configuration file, entity can be automatically generated!\n`));
                     context.useConfigurationFile = true;
                 }
+
+                context.entitySuffix = configuration.get('entitySuffix');
+                if (context.entitySuffix === undefined || context.entitySuffix === null) context.entitySuffix = '';
```
Comment (MathieuAA) [inline generators/entity/index.js]:
```javascript
if (!context.entitySuffix) {
   context.entitySuffix = '';
}
```

File: generators/entity/index.js [hunk 8]
```diff
@@ -223,6 +223,18 @@ module.exports = class extends BaseGenerator {
...
+                if (context.entitySuffix === false) context.entitySuffix = '';
+
+                context.dtoSuffix = configuration.get('dtoSuffix');
+                if (context.dtoSuffix === undefined || context.dtoSuffix === null) context.dtoSuffix = 'DTO';
+                if (context.dtoSuffix === false) context.dtoSuffix = '';
```
Comment (MathieuAA) [inline generators/entity/index.js]: This line can be removed

File: generators/app/index.js (hunk 6 above)
Comment (mselerin) [reply generators/app/index.js]: OK, just to be sure... Only the `false` check is not required anymore ?
The fix has been provided inside jhipster-core ?

**Comment by MathieuAA** [comment]
Yep :)

File: generators/app/index.js (hunk 6 above)
Comment (mselerin) [reply generators/app/index.js]: Just to let you know...

I can't use this in Javascript : `this.config.get('dtoSuffix') || this.options['dto-suffix']` because if `this.config.get('dtoSuffix')` is an empty string, it's a falsy value, thus the `this.options['dto-suffix']` is applied.

I need to do this : 
```javascript
this.dtoSuffix = this.configOptions.dtoSuffix = _.isNil(this.config.get('dtoSuffix'))
            ? this.options['dto-suffix']
            : this.config.get('dtoSuffix');
```
(`_.isNil` is true if the value is null or undefined)

File: generators/entity/index.js (hunk 7 above)
Comment (mselerin) [reply generators/entity/index.js]: Same as my previous comment, I need to do this : 
```javascript
if (_.isNil(context.dtoSuffix)) {
   context.dtoSuffix = 'DTO';
}
```

**Comment by mselerin** [comment]
@MathieuAA all changes are done :-)

**Comment by MathieuAA** [comment]
Awesome

**Comment by deepu105** [comment]
shit seems like we merged the docs but not the PR

**Comment by deepu105** [comment]
@jhipster/developers can someone test this please

**Comment by deepu105** [comment]
@mselerin could you enable this in one of the travis test entity?

**Comment by mselerin** [comment]
@deepu105 OK, I'll update the tests ASAP.

BTW, I've merge the master in the PR to keep it up-to-date ;-)

**Comment by mselerin** [comment]
I have detected another error when generating a new app with entity suffix and an entity linked to the `User` entity.

The User entity is generated with the `server` sub-generator and doesn't respect the suffixes. So, when a new entity has a relation with User, it assumes that the class name would be `UserEntity` (if the entity suffix is 'Entity').

@deepu105, what do you think ? Should I modify the server generator as well ?

**Comment by mselerin** [comment]
I have resolved the conflict with master (but messed on my first attempt, so had to rewrite history on my branch, :/)

**Comment by ruddell** [comment]
Tested a few app combinations and the CI passed, so merging.  Congrats and nice work 👍
